sdk/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java

```diff
@@ -0,0 +1,70 @@
+package io.opentelemetry.sdk.common.export;
+
+import java.util.concurrent.Executor;
+
+/**
+ * The implementation of Export operations are often asynchronous in nature, hence the need to
+ * convey a result at a later time. CompletableResultCode facilitates this.
+ *
+ * <p>This class models JDK 8's CompletableFuture to afford migration should Open Telemetry's SDK
+ * select JDK 8 or greater as a baseline, and also to offer familiarity to developers.
+ */
+public class CompletableResultCode {
+  public CompletableResultCode() {}
+
+  private volatile boolean succeeded = false;
+  private boolean completed = false;
+  private Runnable action = null;
+  private Executor actionExecutor = null;
+
+  /** The export operation finished successfully. */
+  public synchronized void success() {
+    if (!completed) {
+      if (!succeeded) {
+        succeeded = true;
+      }
+      if (action != null) {
+        actionExecutor.execute(action);
+      }
+      completed = true;
+    }
+  }
+
+  /** The export operation finished with an error. */
+  public synchronized void failure() {
+    if (!completed) {
+      if (action != null) {
+        actionExecutor.execute(action);
+      }
+      completed = true;
+    }
+  }
+
+  /**
+   * Obtain the current state of completion. Generally call once completion is achieved via the
+   * thenRun method.
+   *
+   * @return the current state of completion
+   */
+  public boolean isSuccess() {
+    return succeeded;
+  }
+
+  /**
+   * Perform an action on completion. Actions are guaranteed to be called only once.
+   *
+   * <p>There should only be one action for this class instance.
```

Comment:
If we only support one callback anyways I would definitely prefer just using the callback pattern. How about we just go ahead and maintain a list of runnables here?

Reply:
I’d prefer to avoid what would be additional complexity, but composing multiple `andThenAsync` methods would be the idiomatic thing to do.

Reply:
I think we're sort of not past the cliff in terms of complexity. For comparison, the callback pattern is just defining an interface, so it's about as close to no-complexity as changes can get. Here we've added a lot of complexity, having to worry about synchronization to boot, but it seems the only additional feature is a custom executor, which is arguably not that important (our goal was to make the API usable from an async context). So unless we go a bit further, I don't see any reason to take this approach over a normal callback.

Thinking again (thanks for bringing up the point about complexity, I was forgetting it :) ), if we can only imagine a use case where there is a single callback that does something very simple like do a log statement, update a metric, then I'd still lean towards the callback pattern (but I'm still ok with the listenable `Completable` as well if others prefer it). 

It does bring to mind another idea, what about dropping the result code completely? Just a single code seems quite limiting anyways, instead registering a listener on the exporter itself (not on export calls) seems like it could cover all expected use cases, possibly in a more controlled fashion? For example, Zipkin's reporter metrics callback could work well if registered directly on the `Exporter`, and allows consistency in handling success, failure, etc across all exporters.

https://github.com/openzipkin/zipkin-reporter-java/blob/master/core/src/main/java/zipkin2/reporter/ReporterMetrics.java#L44

Reply:
I see have a completable status as useful in that the sdk should provide telemetry on a best-effort basis. If a previous export has not yet finished, then we should look at dropping the metrics. I shall include this change so we can then debate its merit.

In terms of complexity, `andThen`, when used with JDK 8's `CompletableFuture` requires the guarantee that any subsequently composed `andThen` occurs after. Therefore, we would have to maintain a list of actions. I can do this.